File: 0599-minimum-index-sum-of-two-lists/0599-minimum-index-sum-of-two-lists.cs
public class Solution {
    public string[] FindRestaurant(string[] list1, string[] list2) {
        //  Editorial
        //  1.  Using HashMap
        // Dictionary<int, List<string>> dic = new();
        // for(int i = 0; i < list1.Length; i++){
        //     for(int j = 0; j < list2.Length; j++){
        //         if(list1[i] == list2[j]){
        //             if(!dic.ContainsKey(i + j))
        //                 dic.Add(i+j, new List<string>());
        //             dic[i+j].Add(list1[i]);
        //         }
        //     }
        // }
        // int min_index_sum = Int32.MaxValue;
        // foreach(KeyValuePair<int, List<string>> p in dic){
        //     min_index_sum = Math.Min(min_index_sum, p.Key);
        // }
        // return dic[min_index_sum].ToArray();

        //  2.  Without Using HashMap
        // List<string> list = new();
        // for(int sum = 0; sum < list1.Length + list2.Length - 1; sum++){
        //     for(int i = 0; i <= sum; i++){
        //         if(i < list1.Length && sum - i < list2.Length && list1[i] == list2[sum - i])
        //             list.Add(list1[i]);
        //     }
        //     if(list.Count > 0)
        //         break;
        // }
        // return list.ToArray();

        //  3.  Using HashMap(liner)
        Dictionary<string, int> dic = new();
        for(int i = 0; i < list1.Length; i++){
            dic.Add(list1[i],i);
        }
        List<string> list = new();
        int min_sum = Int32.MaxValue, sum;
        for(int j = 0; j < list2.Length && j <= min_sum; j++){
            if(dic.ContainsKey(list2[j])){
                sum = j + dic[list2[j]];
                if(sum < min_sum){
                    list.Clear();
                    list.Add(list2[j]);
                    min_sum = sum;
                }else if(sum == min_sum){
                    list.Add(list2[j]);
                }
            }
        }
        return list.ToArray();
    }
}

File: 0599-minimum-index-sum-of-two-lists/0599-minimum-index-sum-of-two-lists.cpp
class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        map<string, int> mp;
        multimap<int, string> mp2;
        vector<string> ans;
        for(int i = 0; i < list1.size(); i++){
            mp.insert({list1[i], i});
        }

        for(int i = 0; i < list2.size(); i++){
            if(mp.contains(list2[i])){
                mp2.insert({mp[list2[i]] + i, list2[i]});
            }
        }
        int minIndex = -1;
        for(auto p : mp2){
            if(minIndex == -1){
                minIndex = p.first;
                ans.push_back(p.second);
            }else if(minIndex == p.first){
                ans.push_back(p.second);
            }else{
                break;
            }
        }
        return ans;

        //  ==============================  //
        //  Editorial도 참고

        //  https://leetcode.com/problems/minimum-index-sum-of-two-lists/solutions/103674/c-9-lines-hash-table-easy-to-understand
        // vector<string>res;
        // unordered_map<string,int>m;
        // int min = INT_MAX;
        // for(int i = 0; i < list1.size(); i++) m[list1[i]] = i;
        // for(int i = 0; i < list2.size(); i++)
        //     if(m.count(list2[i]) != 0)
        //         if(m[list2[i]] + i < min) min = m[list2[i]] + i, res.clear(), res.push_back(list2[i]);
        //         else if(m[list2[i]] + i == min) res.push_back(list2[i]);
        // return res;
    }
};